### DISCUSSION TITLE
Custom Button Windows

### DISCUSSION THREAD

**Discussion by ghost** [MahApps/MahApps.Metro#4034] [Q&A] [answered]
How do I add color to MetroWindows buttons?
Mine are gray.

```
    <Style x:Key="MeuMetroWindow" TargetType="{x:Type mah:MetroWindow}">
        <Setter Property="NonActiveWindowTitleBrush" Value="{DynamicResource BlueBrush}" />
        <Setter Property="TitleAlignment" Value="Center" />
        <Setter Property="TitleForeground" Value="{DynamicResource WhiteBrush}"/>
        <Setter Property="Background"  Value="{DynamicResource WhiteBrush}"/>
        <Setter Property="BorderBrush"  Value="{DynamicResource BlueBrush}" />
        <Setter Property="WindowTitleBrush" Value="{DynamicResource BlueBrush}" />
    </Style>
```

**Comment by timunie** [comment]
Hi @fabiociaf ,

I think there are two issues with your `Style`. 
1. You should use either `BasedOn="{StaticResource {x:Type mah:MetroWindow}` 
2. I think `BlueBrush` and `WhiteBrush`  are the wrong keys. 

if you are on MahApps v2.0 it should be: 
```xaml
<Style x:Key="MeuMetroWindow" TargetType="{x:Type mah:MetroWindow}" BasedOn="{StaticResource {x:Type mah:MetroWindow}>
    <Setter Property="NonActiveWindowTitleBrush" Value="{DynamicResource MahApps.Brushes.Accent}" />
    <Setter Property="TitleAlignment" Value="Center" />
    <Setter Property="TitleForeground" Value="{DynamicResource MahApps.Brushes.IdealForeground}"/>
    <Setter Property="Background"  Value="{DynamicResource MahApps.Brushes.ThemeBackground}"/>
    <Setter Property="BorderBrush"  Value="{DynamicResource MahApps.Brushes.Accent}" />
    <Setter Property="WindowTitleBrush" Value="{DynamicResource MahApps.Brushes.Accent}" />
</Style>
```

if you want to override the `Buttons` itself you may override these `Styles` with the same `x:Key`: 
https://github.com/MahApps/MahApps.Metro/blob/bbab412af829c9b565fdfdbfd7aeacd9a30805f3/src/MahApps.Metro/Styles/Controls.Buttons.xaml#L313-L383

Happy coding
Tim

**Reply by ghost** [reply]
Hi Tim, 
I made changes to Controls.Buttons in my project. However I need to link with my change and not with the MahAppMetro that he is getting.
I am using the new version v2.4.3
The `BlueBrush `and `WhiteBrush `are my colors of blue and white that have to go but put in the project it is coming gray.

**Reply by timunie** [reply]
@fabiociaf could you upload a short sample App with your custom Brushes and Styles to github and share the link here? Right now I have no idea what might be wrong.

**Reply by ghost** [reply]
Tim, 
My window looks just like I show in these two images.
I have to use my blue and my white.

![image](https://user-images.githubusercontent.com/26013017/106605664-e7215b00-653f-11eb-9f7d-67fa5fbe0600.png)
![image](https://user-images.githubusercontent.com/26013017/106605823-1afc8080-6540-11eb-8add-03fdd59660fe.png)

**Reply by timunie** [reply]
Understood, but without a look into your code it is hard to guess what is wrong. In case you use `BlueBrush` and `WhiteBrush` everywhere I would consider using a custom `Theme`

Here is a guide: https://github.com/ControlzEx/ControlzEx/blob/develop/Wiki/ThemeManager.md 

Maybe you can upload an App which shows this and than we may inspect it further.

**Reply by ghost** [reply]
I no use `Theme`

My code:

TELACADUSUARIOS.xaml (MetroWindow)

```
<mah:MetroWindow 
    x:Class="DFe.TELACADUSUARIOS"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:mah="http://metro.mahapps.com/winfx/xaml/controls"  
    xmlns:local="clr-namespace:DFe"
    xmlns:local2="clr-namespace:DFe.Entidades"
    xmlns:fa="http://schemas.fontawesome.com/icons/"
    KeyDown="MetroWindow_KeyDown"
    mc:Ignorable="d"
    Style="{StaticResource MeuMetroWindowJanelas}"        
    xmlns:collections="clr-namespace:System.Collections;assembly=mscorlib"
    d:DataContext="{d:DesignInstance {x:Type local:Usuarios}, IsDesignTimeCreatable=True}"
    Background="{StaticResource CordeFundo}"
    WindowStartupLocation="CenterScreen"
    Loaded="MetroWindow_Loaded"
    Height="500" Width="540"
    Title="USUÁRIOS E PERMISSÕES DE ACESSO" >
<Grid>....</Grid></mah:MetroWindow>
```

And  Window.xaml (style)

```
<ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:mah="http://metro.mahapps.com/winfx/xaml/controls" 
                     xmlns:local="clr-namespace:DFe" >
<Style x:Key="MeuMetroWindow" TargetType="{x:Type mah:MetroWindow}">
        <Setter Property="NonActiveWindowTitleBrush" Value="{DynamicResource BlueBrush}" />
        <Setter Property="TitleAlignment" Value="Center" />
        <Setter Property="TitleForeground" Value="{DynamicResource WhiteBrush}"/>
        <Setter Property="Background"  Value="{DynamicResource WhiteBrush}"/>
        <Setter Property="BorderBrush"  Value="{DynamicResource BlueBrush}" />
        <Setter Property="WindowTitleBrush" Value="{DynamicResource BlueBrush}" />
    </Style>
</ResourceDictionary>

```

**Reply by timunie** [reply]
Please 🙏 create a new repro and upload a sample which can be run directly. I have no time to reimplement all the stuff. Thx

**Reply by ghost** [reply]
@timunie  
https://github.com/fabiociaf/WpfCustomButtonWindows

**Comment by timunie** [comment, marked as answer]
Hi @fabiociaf 

The issue is that you do not import all the needed styles. 

in your `App.xaml` you need to add these to `MergedDictionaries`:
```xaml
<!--  MahApps.Metro resource dictionaries. Make sure that all file names are Case Sensitive!  -->
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Controls.xaml" />
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Fonts.xaml" />
<!--  Theme setting  -->
<ResourceDictionary Source="pack://application:,,,/MahApps.Metro;component/Styles/Themes/Light.Blue.xaml" />
```

See: https://mahapps.com/docs/guides/quick-start

And here is the result: 
![image](https://user-images.githubusercontent.com/47110241/106631533-2621d180-657d-11eb-8afd-1844ed76948e.png)

If you don't want to have MahApps-Style then please read: #4029

I'll send a PR to your sample

Happy coding
Tim

**Reply by ghost** [reply]
Tim,
You decided with a theme I will have to accept.
You helped me a lot I will need to review the theme part to start using.
Thanks you!

**Comment by ghost** [comment]
@timunie 
I have new issue. I have new windows but now with the colors switched. What was blue turns white and what was white turns blue.


![image](https://user-images.githubusercontent.com/26013017/106638224-66c02180-6562-11eb-97c8-b30ced933189.png)

That's why I didn't want to use `Theme`


```
<Style x:Key="MeuMetroWindow2" TargetType="{x:Type mah:MetroWindow}" BasedOn="{StaticResource {x:Type mah:MetroWindow}}">
        <Setter Property="NonActiveWindowTitleBrush" Value="{DynamicResource BlueBrush}" />
        <Setter Property="TitleAlignment" Value="Center" />
        <Setter Property="TitleForeground" Value="{DynamicResource BlueBrush}" />
        <Setter Property="Background"  Value="{StaticResource WhiteBrush}"/>
        <Setter Property="BorderBrush"  Value="{DynamicResource BlueBrush}" />
        <Setter Property="WindowTitleBrush" Value="{StaticResource WhiteBrush}"/>
        <Setter Property="Foreground" Value="{DynamicResource BlueBrush}" />
    </Style>
```

**Reply by timunie** [reply]
Hi @fabiociaf 

I am sorry, I didn't consider this scenario. Here you will need to implement a custom `Style` for the `WindowButtonCommands`

```xaml 
<Style TargetType="mah:WindowButtonCommands" BasedOn="{StaticResource MahApps.Styles.WindowButtonCommands.Base}">
    <Setter Property="Foreground" Value="{Binding RelativeSource={RelativeSource Mode=FindAncestor, AncestorType=mah:MetroWindow}, Path=TitleForeground}" />
    <Setter Property="Template" Value="{DynamicResource MahApps.Templates.WindowButtonCommands.Win10}" />
</Style>
```

I made another PR to your sample.

@punker76 should we add this Setter `<Setter Property="Foreground" Value="{Binding RelativeSource={RelativeSource Mode=FindAncestor, AncestorType=mah:MetroWindow}, Path=TitleForeground}" />` to the original `Style`? I think it is the right way to inherit the `TitleForeground` here. 

Happy coding
Tim

**Reply by ghost** [reply]
Thanks my friends 
I love it.
